src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/CockroachDBEngine.java

```diff
@@ -0,0 +1,384 @@
+import com.feedzai.commons.sql.abstraction.ddl.DbColumnType;
+import com.feedzai.commons.sql.abstraction.ddl.DbEntity;
+import com.feedzai.commons.sql.abstraction.dml.result.ResultColumn;
+import com.feedzai.commons.sql.abstraction.engine.AbstractTranslator;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngineDriver;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngineException;
+import com.feedzai.commons.sql.abstraction.engine.MappedEntity;
+import com.feedzai.commons.sql.abstraction.engine.configuration.PdbProperties;
+import com.feedzai.commons.sql.abstraction.engine.handler.OperationFault;
+import com.feedzai.commons.sql.abstraction.entry.EntityEntry;
+
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.feedzai.commons.sql.abstraction.util.StringUtils.md5;
+import static com.feedzai.commons.sql.abstraction.util.StringUtils.quotize;
+import static java.lang.String.format;
+import static org.apache.commons.lang3.StringUtils.join;
+
+/**
+ * CockroachDB specific database implementation.
+ *
+ * @author Mário Pereira (mario.arzileiro@feedzai.com)
+ * @since 2.5.0
+ */
+public class CockroachDBEngine extends PostgreSqlEngine {
+
+    /**
+     * The Cockroach JDBC (PostgreSQL) driver.
+     */
+    protected static final String COCKROACHDB_DRIVER = DatabaseEngineDriver.COCKROACHDB.driver();
+
+    /**
+     * Constraint name already exists.
+     *
+     * Note: CockroachDB v19.2.x returns this SQL state when trying to add a foreign key relation that already exists.
+     * This constant is needed because PosgtgreSQL uses code 42710 - ERRCODE_DUPLICATE_OBJECT instead.
+     * In this situation CockroachDB v19.2.x uses code 23503 - ERRCODE_FOREIGN_KEY_VIOLATION (a column in a table has
+     * a value that is not present in the column of other table in the foreign key relation);
+     * in the same situation CockroachDB v19.1.x returns state 42830 - ERRCODE_INVALID_FOREIGN_KEY (which can occur for
+     * example when the types of columns in the relation have incompatible types).
+     * @see <a href="https://github.com/cockroachdb/cockroach/issues/42858">CockroachDB issue #42858</a>
+     */
+    public static final String CONSTRAINT_NAME_ALREADY_EXISTS_COCKROACH = "23503";
+
+    /**
+     * Creates a new CockroachDB connection.
+     *
+     * @param properties The properties for the database connection.
+     * @throws DatabaseEngineException When the connection fails.
+     */
+    public CockroachDBEngine(final PdbProperties properties) throws DatabaseEngineException {
+        super(properties, COCKROACHDB_DRIVER);
+    }
+
+    @Override
+    protected void createTable(final DbEntity entity) throws DatabaseEngineException {
```

Comment:
There's a lot of code here that is duplicate of the one in the Posgres engine and that could be eliminated with a minor refactor. Doing that would help to understand what is different between Posgres and Cockroach.

Reply:
You're correct - and it is planned (issue #126).
CockroachDB needs to have a primary key created along with the table, otherwise a default one is created, and afterwards it is not possible to add another one or replace the default.
The other engines don't need it, but they can also have a CREATE TABLE statement that also defines primary key, indexes, etc; when this is changed for PostgreSQL, then the override for CockroachDB probably can be removed.

Reply:
Please ensure that is documented in issue #126 then, didn't see any reference to this.

Reply:
It's on point 5.
#126 is a bit messy, each of those things should have it's own ticket#, but from the introduction of CockroachDB these issues kept coming up and it was easier to just add more to an existing list than creating a new issue for each one :smile_cat: